src/core/lib/surface/server.cc

```diff
@@ -1211,30 +1171,21 @@ void Server::ChannelData::InitTransport(RefCountedPtr<Server> server,
   grpc_transport_perform_op(transport, op);
 }
 
-Server::ChannelRegisteredMethod* Server::ChannelData::GetRegisteredMethod(
-    const grpc_slice& host, const grpc_slice& path) {
-  if (registered_methods_ == nullptr) return nullptr;
-  // TODO(ctiller): unify these two searches
-  // check for an exact match with host
-  uint32_t hash = MixHash32(grpc_slice_hash(host), grpc_slice_hash(path));
-  for (size_t i = 0; i <= registered_method_max_probes_; i++) {
-    ChannelRegisteredMethod* rm =
-        &(*registered_methods_)[(hash + i) % registered_methods_->size()];
-    if (rm->server_registered_method == nullptr) break;
-    if (!rm->has_host) continue;
-    if (rm->host != host) continue;
-    if (rm->method != path) continue;
-    return rm;
-  }
-  // check for a wildcard method definition (no host set)
-  hash = MixHash32(0, grpc_slice_hash(path));
-  for (size_t i = 0; i <= registered_method_max_probes_; i++) {
-    ChannelRegisteredMethod* rm =
-        &(*registered_methods_)[(hash + i) % registered_methods_->size()];
-    if (rm->server_registered_method == nullptr) break;
-    if (rm->has_host) continue;
-    if (rm->method != path) continue;
-    return rm;
+Server::RegisteredMethod* Server::ChannelData::GetRegisteredMethod(
+    const absl::string_view& host, const absl::string_view& path)
+    ABSL_LOCKS_EXCLUDED(registered_methods_mu_) {
+  MutexLock lock(&server_->registered_methods_mu_);
```

Comment:
We'll need to figure out a way to avoid this lock: since it's on the per-request path it'll be a huge contention point.

One thing to consider is making `registered_methods_` a `shared_ptr<const flat_hash_map<...>>` - if we did that then when we register a new method we could acquire the lock, make a new `shared_ptr<const flat_hash_map<>>` and then install that into registered_methods_ -- and we could then have each channel pick up its own reference to the map.

The effect would be that newly registered methods *after* server instantiation would not be visible to channels, and maybe that's ok.

Another approach would be to crash on registering a method after the server has started -- if we did that, then we don't need a mutex to *read* from registered_methods_ because it's guaranteed immutable at that point.

Reply:
Took approach of crashing if we register after server started.